Emby.Server.Implementations/Session/SessionManager.cs

```diff
@@ -1833,5 +1778,51 @@ public Task SendMessageToUserDeviceSessions<T>(string deviceId, SessionMessageTy
 
             return SendMessageToSessions(sessions, name, data, cancellationToken);
         }
+
+        /// <inheritdoc />
+        public async ValueTask DisposeAsync()
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            foreach (var session in _activeConnections.Values)
+            {
+                await session.DisposeAsync().ConfigureAwait(false);
+            }
+
+            if (_idleTimer is not null)
+            {
+                await _idleTimer.DisposeAsync().ConfigureAwait(false);
+                _idleTimer = null;
+            }
+
+            await _shutdownCallback.DisposeAsync().ConfigureAwait(false);
+
+            _deviceManager.DeviceOptionsUpdated -= OnDeviceManagerDeviceOptionsUpdated;
+            _disposed = true;
+        }
+
+        private async void OnApplicationStopping()
+        {
+            _logger.LogInformation("Sending shutdown notifications");
+            try
+            {
+                var messageType = _appHost.ShouldRestart ? SessionMessageType.ServerRestarting : SessionMessageType.ServerShuttingDown;
+
+                await SendMessageToSessions(Sessions, messageType, string.Empty, CancellationToken.None).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Error sending server shutdown notifications");
+            }
+
+            // Close open websockets to allow Kestrel to shut down cleanly
+            foreach (var session in _activeConnections.Values)
+            {
+                await session.DisposeAsync().ConfigureAwait(false);
```

Comment:
Aren't these disposed twice? Once here and then again in DisposeAsync?

Reply:
Yes, kestrel waits for the web sockets to close before it is considered to have shut down, so it would hang if this wasn't here. This doesn't cause any problems, since disposing something a second time has no effect (or shouldn't, anyway), but ideally we would do something like passing the ApplicationStopping token to the web sockets so that they stop without needing this, but idk if that would introduce a race with sending out the shutdown messages, since they also happen on that event.